app/uk/gov/hmrc/personaldetailsvalidation/PersonalDetailsValidator.scala

```diff
@@ -52,16 +52,28 @@ private class PersonalDetailsValidator[Interpretation[_] : Monad](matchingConnec
   def validate(personalDetails: PersonalDetails)
               (implicit hc: HeaderCarrier, request: Request[_], ec: ExecutionContext): EitherT[Interpretation, Exception, ValidationId] = {
     for {
-      matchResult <- doMatch(personalDetails)
+      correctContent <- validateContentRules(personalDetails)
+      matchResult <- doMatch(correctContent)
       personalDetailsValidation = matchResult.toPersonalDetailsValidation(optionallyHaving = personalDetails)
       _ <- personalDetailsValidationRepository.create(personalDetailsValidation)
       _ = sendEvents(matchResult, personalDetails)
     } yield personalDetailsValidation.id
-  }.leftMap { error => sendErrorEvents(personalDetails); error }
+  }.leftMap {
+    error => sendErrorEvents(personalDetails); error
+  }
+
+  private def validateContentRules(personalDetails: PersonalDetails): EitherT[Interpretation, Exception, PersonalDetails] = {
+    personalDetails match {
+      case bad if (bad.nino.isDefined && bad.postCode.isDefined) => EitherT.leftT[Interpretation, PersonalDetails](new IllegalArgumentException("both nino and postcode supplied"))
```

Comment:
This validation can move into Reads

Reply:
I agree with Anshul. This will take lines of code from here and allow to use Play mechanisms to turn form validation errors to BAD_REQUEST. Actually, the logic of extracting domain objects from request's payload belongs to the form binding and it seems reasonable that it does not leak out of the page PersonalDetailsPage.